### PR TITLE
[Debuginfo][GVNHoist] Fix #86227: update the debug location of the hoisted GEP

### LINKED ISSUE #86227: [Debuginfo][GVNHoist] The hoisted GEP preserves the debug location from one of the conditional branches
When hoisting GEPs used by hoisted instructions from `if.then` and `if.else`, the debug location of the hoisted GEP is not properly updated, preserving the debug location of the cloned GEP, ie., the GEP in `if.then` in the following program. However, the debug location of the hoisted GEP should be the merge of those in two branches, ie., applying a merging update. 

```LLVM
; Function Attrs: noinline nounwind uwtable
define dso_local void @func(i32 noundef %a, ptr noundef %b) #0 !dbg !10 {
entry:
  tail call void @llvm.dbg.value(metadata i32 %a, metadata !16, metadata !DIExpression()), !dbg !17
  tail call void @llvm.dbg.value(metadata ptr %b, metadata !18, metadata !DIExpression()), !dbg !17
  %tobool = icmp ne i32 %a, 0, !dbg !19
  br i1 %tobool, label %if.then, label %if.else, !dbg !21

if.then:                                          ; preds = %entry
  %arrayidx = getelementptr inbounds i32, ptr %b, i64 1, !dbg !22
  store i32 1, ptr %arrayidx, align 4, !dbg !24
  br label %if.end, !dbg !25

if.else:                                          ; preds = %entry
  %arrayidx1 = getelementptr inbounds i32, ptr %b, i64 1, !dbg !26
  store i32 1, ptr %arrayidx1, align 4, !dbg !28
  br label %if.end

if.end:                                           ; preds = %if.else, %if.then
  ret void, !dbg !29
}

...
!22 = !DILocation(line: 3, column: 9, scope: !23)
...
!26 = !DILocation(line: 5, column: 9, scope: !27)
...
```

```LLVM
; ModuleID = 'main.ll'
source_filename = "main.c"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

; Function Attrs: noinline nounwind uwtable
define dso_local void @func(i32 noundef %a, ptr noundef %b) #0 !dbg !10 {
entry:
  tail call void @llvm.dbg.value(metadata i32 %a, metadata !16, metadata !DIExpression()), !dbg !17
  tail call void @llvm.dbg.value(metadata ptr %b, metadata !18, metadata !DIExpression()), !dbg !17
  %tobool = icmp ne i32 %a, 0, !dbg !19
  %0 = getelementptr inbounds i32, ptr %b, i64 1, !dbg !21
  store i32 1, ptr %0, align 4
  br i1 %tobool, label %if.then, label %if.else, !dbg !23

if.then:                                          ; preds = %entry
  %arrayidx = getelementptr inbounds i32, ptr %b, i64 1, !dbg !21
  br label %if.end, !dbg !24

if.else:                                          ; preds = %entry
  %arrayidx1 = getelementptr inbounds i32, ptr %b, i64 1, !dbg !25
  br label %if.end

if.end:                                           ; preds = %if.else, %if.then
  ret void, !dbg !27
}

...
!21 = !DILocation(line: 3, column: 9, scope: !22)
...
!25 = !DILocation(line: 5, column: 9, scope: !26)
...
```

The [hoisting code](https://github.com/llvm/llvm-project/blob/a20f7efbc587a213868b494d3d34a8cbeaff04ab/llvm/lib/Transforms/Scalar/GVNHoist.cpp#L924):
```C++
void GVNHoist::makeGepsAvailable(Instruction *Repl, BasicBlock *HoistPt,
                                 const SmallVecInsn &InstructionsToHoist,
                                 Instruction *Gep) const {
  assert(allGepOperandsAvailable(Gep, HoistPt) && "GEP operands not available");

  Instruction *ClonedGep = Gep->clone();
  ...

  // Copy Gep and replace its uses in Repl with ClonedGep.
  ClonedGep->insertBefore(HoistPt->getTerminator());

  ...

  // Replace uses of Gep with ClonedGep in Repl.
  Repl->replaceUsesOfWith(Gep, ClonedGep);
}
```

By the way, the debug location of Its user `store` was updated by dropping (the [commit](https://github.com/llvm/llvm-project/commit/368681f803830f42c539fe37e753d8401a121454)). However, I think merging would be a better choice, because the hoisted `store` is the merge of the two StoreInst in two branches. Merging could preserve the scope information of the merged `store`.

### PR DESCRIPTION
**Pull request by Apochens**
This PR fixes #86227.

### CODE REVIEW

**Comment by Apochens** [comment]
@jmorse @SLTozer @dwblaikie  Please review this.

File: llvm/test/Transforms/GVNHoist/hoist-merge-geps.ll [hunk 1]
```diff
@@ -0,0 +1,66 @@
+; RUN: opt -S -passes=gvn-hoist < %s | FileCheck %s 
+
+; Function Attrs: noinline nounwind uwtable
```
Comment (SLTozer) [inline llvm/test/Transforms/GVNHoist/hoist-merge-geps.ll:3]: Minor nits, we can delete the `; Function Attrs` lines and the attribute references (`#0` and `#1`).

**Review by SLTozer — COMMENTED** [review]
Change and test LGTM, with some minor nits for the test, and also `clang-format` is needed (you can see the suggested changes from the PR checks [here](https://github.com/llvm/llvm-project/actions/runs/8384940424/job/22963060233?pr=86236)).

File: llvm/test/Transforms/GVNHoist/hoist-merge-geps.ll (hunk 1 above)
Comment (Apochens) [reply llvm/test/Transforms/GVNHoist/hoist-merge-geps.ll]: Thanks for your comments. I fixed these minor issues, and now the test would be better.

**Review by SLTozer — APPROVED** [review]
Changes look good!

**Comment by Apochens** [comment]
Ping!

**Pull request merged**